firebase_admin/messaging.py

```diff
@@ -245,25 +359,33 @@ def encode_message(cls, message):
         return cls.JSON_ENCODER.default(message)
 
     def send(self, message, dry_run=False):
-        data = {'message': _MessagingService.encode_message(message)}
-        if dry_run:
-            data['validate_only'] = True
+        request = self._message_request(message, dry_run)
         try:
-            headers = {
-                'X-GOOG-API-FORMAT-VERSION': '2',
-                'X-FIREBASE-CLIENT': self._client_version,
-            }
-            resp = self._client.body(
-                'post', url=self._fcm_url, headers=headers, json=data, timeout=self._timeout)
-        except requests.exceptions.RequestException as error:
-            if error.response is not None:
-                self._handle_fcm_error(error)
-            else:
-                msg = 'Failed to call messaging API: {0}'.format(error)
-                raise ApiCallError(self.INTERNAL_ERROR, msg, error)
+            resp = request.execute(num_retries=4)
+        except googleapiclient.errors.HttpError as error:
+            raise _MessagingService._parse_fcm_error(error)
         else:
             return resp['name']
 
+    def send_all(self, messages, dry_run=False):
+        message_count = len(messages)
+        send_all_complete = threading.Event()
+        responses = []
+
+        def send_all_callback(request_id, response, exception):
+            send_response = SendResponse(response, exception)
+            responses.append(send_response)
+            if len(responses) == message_count:
+                send_all_complete.set()
+
+        batch = self._fcm_service.new_batch_http_request(callback=send_all_callback)
+        for message in messages:
+            batch.add(self._message_request(message, dry_run))
+        batch.execute()
```

Comment:
Batches don't seem to be able to be retried when executing - are we okay with that?